mlflow/server/js/src/experiment-tracking/components/ExperimentListView.tsx

```diff
@@ -25,146 +25,158 @@ import Routes from '../routes';
+  const handleRenameExperiment = (experimentId: any, experimentName: any) => () => {
+    setState((state) => ({
+      ...state,
       showRenameExperimentModal: true,
-    });
-    this.updateSelectedExperiment(experimentId, experimentName);
+    }));
+    updateSelectedExperiment(experimentId, experimentName);
   };
 
-  handleCloseCreateExperimentModal = () => {
-    this.setState({
+  const handleCloseCreateExperimentModal = () => {
+    setState((state) => ({
+      ...state,
       showCreateExperimentModal: false,
-    });
+    }));
   };
 
-  handleCloseDeleteExperimentModal = () => {
-    this.setState({
+  const handleCloseDeleteExperimentModal = () => {
+    setState((state) => ({
+      ...state,
       showDeleteExperimentModal: false,
-    });
+    }));
     // reset
-    this.updateSelectedExperiment('0', '');
+    updateSelectedExperiment('0', '');
   };
 
-  handleCloseRenameExperimentModal = () => {
-    this.setState({
+  const handleCloseRenameExperimentModal = () => {
+    setState((state) => ({
+      ...state,
       showRenameExperimentModal: false,
-    });
+    }));
     // reset
-    this.updateSelectedExperiment('0', '');
+    updateSelectedExperiment('0', '');
+  };
+
+  const pushExperimentRoute = () => {
+    if (state.checkedKeys.length > 0) {
+      const route =
+        state.checkedKeys.length === 1
+          ? Routes.getExperimentPageRoute(state.checkedKeys[0])
+          : Routes.getCompareExperimentsPageRoute(state.checkedKeys);
+      navigate(route);
+    }
   };
 
   // Add a key if it does not exist, remove it if it does
   // Always keep at least one experiment checked if it is only the active one.
-  handleCheck = (isChecked: any, key: any) => {
-    this.setState((prevState: any, props: any) => {
+  const handleCheck = (isChecked: any, key: any) => {
+    setState((prevState) => {
       let { checkedKeys } = prevState;
-      if (isChecked === true && !props.activeExperimentIds.includes(key)) {
-        checkedKeys = [key, ...props.activeExperimentIds];
+      if (isChecked === true && !activeExperimentIds.includes(key)) {
+        checkedKeys = [key, ...activeExperimentIds];
       }
-      if (isChecked === false && props.activeExperimentIds.length !== 1) {
-        checkedKeys = props.activeExperimentIds.filter((i: any) => i !== key);
+      if (isChecked === false && activeExperimentIds.length !== 1) {
+        checkedKeys = activeExperimentIds.filter((i: any) => i !== key);
       }
-      return { checkedKeys: checkedKeys };
-    }, this.pushExperimentRoute);
-  };
-
-  pushExperimentRoute = () => {
-    if (this.state.checkedKeys.length > 0) {
-      const route =
-        this.state.checkedKeys.length === 1
-          ? Routes.getExperimentPageRoute(this.state.checkedKeys[0])
-          : Routes.getCompareExperimentsPageRoute(this.state.checkedKeys);
-      this.props.navigate(route);
-    }
+      const new_state = { ...prevState, checkedKeys: checkedKeys }
+      pushExperimentRoute()
```

Comment:
it seems like in functional react we have to use `useEffect` if we want to guarantee that something gets executed after state updates.

i propose deleting `pushExperimentRoute()` and just inlining a useEffect above like

```typescript
  useEffect(() => {
    if (state.checkedKeys.length > 0) {
      const route =
        state.checkedKeys.length === 1
          ? Routes.getExperimentPageRoute(state.checkedKeys[0])
          : Routes.getCompareExperimentsPageRoute(state.checkedKeys);
      navigate(route);
    }
  }, [state.checkedKeys, navigate]);
  ```
  
  otherwise a bug like this occurs:
  

https://github.com/user-attachments/assets/3f5c29ab-419c-4050-bfca-b42932056645